include/picongpu/fields/background/templates/twtstight/TWTSTight.tpp

```diff
@@ -0,0 +1,366 @@
+            {
+                float_64 const time_SI = float_64(currentStep) * dt - tdelay;
+
+                pmacc::math::Vector<floatD_64, detail::numComponents> const fieldPositions_SI
+                    = detail::getFieldPositions_SI(cellIdx, halfSimSize, extraShifts, unit_length, focus_y_SI, phi);
+
+                /* Single TWTS-Pulse */
+                return getTWTSField_Normalized(fieldPositions_SI, time_SI);
+            }
+
+            template<typename T_Field>
+            template<uint32_t T_component>
+            HDINLINE float_X TWTSTight<T_Field>::getComponent(floatD_X const& cellIdx, float_X const currentStep) const
+            {
+                // The optimized way is only implemented for 3D, fall back to full field calculation in 2d
+                if constexpr(simDim == DIM3)
+                {
+                    float_64 const time_SI = float_64(currentStep) * dt - tdelay;
+                    pmacc::math::Vector<floatD_X, detail::numComponents> zeroShifts;
+                    for(uint32_t component = 0; component < detail::numComponents; ++component)
+                        zeroShifts[component] = floatD_X::create(0.0);
+                    pmacc::math::Vector<floatD_64, detail::numComponents> const fieldPositions_SI
+                        = detail::getFieldPositions_SI(cellIdx, halfSimSize, zeroShifts, unit_length, focus_y_SI, phi);
+                    // Explicitly use a 3D vector so that this function compiles for 2D as well
+                    auto const pos = float3_64{
+                        fieldPositions_SI[T_component][0],
+                        fieldPositions_SI[T_component][1],
+                        fieldPositions_SI[T_component][2]};
+                    if constexpr(T_component == 0)
+                        return static_cast<float_X>(calcTWTSFieldX(pos, time_SI));
+                    else
+                    {
+                        if constexpr(T_component == 1)
+                        {
+                            return static_cast<float_X>(calcTWTSFieldY(pos, time_SI));
+                        }
+                        if constexpr(T_component == 2)
+                        {
+                            return static_cast<float_X>(calcTWTSFieldZ(pos, time_SI));
+                        }
+                    }
+
+                    // we should never be here
+                    return NAN;
+                }
+                if constexpr(simDim != DIM3)
+                    return (*this)(cellIdx, currentStep)[T_component];
+            }
+
+            template<typename T_Field>
+            HDINLINE std::tuple<
+                float_T,
+                float_T,
+                float_T,
+                float_T,
+                float_T,
+                float_T,
+                float_T,
+                float_T,
+                float_T,
+                float_T,
+                float_T>
+            TWTSTight<T_Field>::defineBasicHelperVariables() const
```

Comment:
Managing the correct order of TWTSTight members for calculations within initializer lists is error prone. It makes the number of data members cluttered and documentation less easy to follow. This could in principle be solved with nested structs, but then I cannot use the struct members in the same way as the members of TWTSTight, but have to type something like `structInstance.tanAlpha` which makes consecutive equations much more cluttered. This argument is the reason we are using constructive bindings instead to keep everything lightweight.

Reply:
Initialiser lists are typically preferred but if it's better maintainable to simply call a function in the inside of the constructor I don't think it would hurt tremendously, so initialisation could happen inside the constructor body.

Also, structs can be unpacked in the same way as tuples see [here](https://en.cppreference.com/w/cpp/language/structured_binding). So, you could still unpack the names where you need them even if you have them in a struct. But I kind of see that it defeats the purpose a little.